rfcs/0008-leaderless-dns.md

```diff
@@ -0,0 +1,238 @@
+# Distributed DNS Load Balancing
+
+- Feature Name: Distributed DNS Load Balancing
+- Start Date: 2024-01-17
+- RFC PR: [Kuadrant/architecture#0008](https://github.com/Kuadrant/architecture/pull/55)
+- Issue tracking: [Kuadrant/architecture#0008](https://github.com/Kuadrant/architecture/issues/56)
+
+# Summary
+[summary]: #summary
+
+Enable the DNS Operator to manage DNS for one or more hostnames shared across one or more clusters, where a central "hub" cluster is not required or may not be available, and increase the robustness of the DNS Health checks by executing them from all the available clusters.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, the DNS record controller has limited functionality for a single cluster, and none for multiple clusters without OCM. Having the DNSRecord controller enabled to manage DNS in these situations significantly eases onboarding users to the advantages of the DNSRecord controller.
+
+# Diagrams
+[diagrams]: #diagrams
+
+- [Overview](./0008-distributed-dns-assets/distributed-dns-usecases.jpg)
+- [P2P Health Checks](./0008-distributed-dns-assets/distributed-dns-healthchecks.jpg)
+- [DNS Pruning](./0008-distributed-dns-assets/distributed-dns-pruning.jpg)
+- [DNS Controller](./0008-distributed-dns-assets/distributed-dns-dns-operator.jpg)
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Single Cluster
+In a single cluster scenario, create a DNS Policy that references a DNS provider. When a listener host defined in the gateway is found that matches one of the available zones in the provider, the addresses of that gateway will be created as A or CNAME records for that host in the appropriate zone file following the configuration outlined in the DNS Policy.
+
+### Single Cluster Health Checks
+If health checks are defined in the DNS Policy, these will also be created and used to remove unhealthy IPs from the Zone file, should all health checks fail, then all A records will be published to avoid an NXDOMAIN response.
+
+## Multiple Cluster
+In a multiple cluster scenario, every cluster requires a DNS Policy configured with a provider secret that specifies a unique `multiClusterID`, they should all reference the same zone(s) in the same DNS Provider for the same host. In each cluster, when a matching listener host defined in the gateway is found that matches one of the available zones in the provider the addresses from that gateway will be appended to the zone file - in a non-destructive manner - allowing multiple clusters to all add their gateways addresses to the same zone file, without destroying each other's values.
```

Comment:
What would be the recommended approach for users to take to manage the setting of the multiClusterID across a large number of clusters? For example, if I had a multi-cluster architecture with 10+ clusters in the loop, having to manually set a multiClusterID value across each of these would be a lot of overhead?

Random thought - could a default multiClusterID value be automatically set based on some cluster metadata value that would be unique to that cluster? End users would then have the option to modify this afterwards if they so choose?

Reply:
The setting of a clusterID is also how the user signals to the DNS Operator that this zone is a distributed zone and should be interacted with differently. So we'd either need to change how that signal is sent to the operator, or require the user to set the clusterID. As the clusterID has no purpose in a single-cluster or OCM scenario, it made more sense to me, that this field is the signal that we are in a multi-cluster scenario.

Reply:
Should we rely on ArgoCD to solve this? Isn't there a better way how to resolve this without Argo?

Reply:
argocd is one way to do it. We wont be using argo it is just a way to solve this. Another way is to do it manually. I am sure you could do it with ansible etc (IE the setting of a multiclusterID is an exercise for the user and is somewhat dependant on their tooling)

Reply:
how about leaving generation of clusterid on the (dns) operator? I think it would suffice to use some random uuid whose value space and proper RNG guarantee distinct values each time (widely used mechanism).

Reply:
> The setting of a clusterID is also how the user signals to the DNS Operator that this zone is a distributed zone and should be interacted with differently. So we'd either need to change how that signal is sent to the operator, or require the user to set the clusterID. As the clusterID has no purpose in a single-cluster or OCM scenario, it made more sense to me, that this field is the signal that we are in a multi-cluster scenario.

I would prefer `mutlicluster: true` flag and to have ClusterID generated from metadata if possible, as that seems the easiest to use

Reply:
This is an interesting idea. So remove the clusterid as a field that can be set and replace with just a boolean. The end user can still make a mistake though. If they set multi-cluster to true on one cluster and false on another and those clusters share a host, they will clash I believe

Reply:
ok, I see the value in this alternative. We generate the clusterID always, and the user simply sets multicluster to true/false.

This makes sense to me, I can update to function this way.

Reply:
It seems odd to add this ID to the provider, it's not really anything to do with the provider, and you will have to add it it to every one you create. This ID should be unique to the controller instance, and probably something you can either set manually on controller init (startup flag) or if possible, something unique you can consistently generate on cluster, no?

Reply:
If we are generating it, I think it can be at the controller level. Not sure if I see any value in allowing it to be set by the user at all now. We prob just need to figure out where we want to store it

Reply:
> It seems odd to add this ID to the provider, it's not really anything to do with the provider, and you will have to add it it to every one you create. This ID should be unique to the controller instance, and probably something you can either set manually on controller init (startup flag) or if possible, something unique you can consistently generate on cluster, no?

I went that way originally, and I think with this change to it being calculated somehow on the cluster and invisible to the user, it will likely return to that system.

Reply:
I think this also means we do not need to manage when the clusterID changes, which is nice as that was feeling like an area with complexity.

Reply:
Updated the RFC now.